### PR TITLE
feat(models): add registries models

### LINKED ISSUE #5019: registries: create pydantic models
### What needs to get done

Formalize pydantic models prototyped in #4917 (types, docstrings, constraints, unit tests).

### Why it needs to get done

To support registries

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/mastex/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Adds pydantic models for registries.

Fixes #5019
(CRAFT-3332)

### CODE REVIEW

**Review by lengau — APPROVED** [review]

File: snapcraft/models/assertions.py [hunk 1]
```diff
@@ -14,29 +14,81 @@
...
+    """Length of the body field."""
+
+    sign_key_sha3_384: str | None = None
+    """Signing key ID."""
+
+
+class RegistriesList(models.CraftBaseModel):
+    """A list of registry assertions."""
+
+    registry_list: list[RegistryAssertion] = []
```
Comment (dariuszd21) [inline snapcraft/models/assertions.py:85]:
```suggestion
    registry_list: list[RegistryAssertion] = Field(default_factory=list)
```

File: tests/unit/models/test_assertions.py [hunk 2]
```diff
@@ -17,11 +17,92 @@
...
 
-def test_assertion_defaults(fake_registry_assertion_data, check):
+
+def test_registry_defaults(check):
+    """Test default values of the Registry model."""
+    registry = Registry.unmarshal({"storage": "test-storage"})
+
+    check.equal(registry.request, None)
+    check.equal(registry.access, None)
+    check.equal(registry.content, None)
```
Comment (dariuszd21) [inline tests/unit/models/test_assertions.py:29]:
```suggestion
    check.is_none(registry.request)
    check.is_none(registry.access)
    check.is_none(registry.content)
```

File: tests/unit/models/test_assertions.py [hunk 3]
```diff
@@ -17,11 +17,92 @@
...
+                        {
+                            "storage": "wifi.ssids",
+                        }
+                    ]
+                }
+            },
+        }
+    )
+
+    check.equal(assertion.summary, None)
```
Comment (dariuszd21) [inline tests/unit/models/test_assertions.py:76]:
```suggestion
    check.is_none(assertion.summary)
```

File: tests/unit/models/test_assertions.py [hunk 4]
```diff
@@ -17,11 +17,92 @@
...
+    check.equal(assertion.revision, 0)
+    check.equal(assertion.body, None)
```
Comment (dariuszd21) [inline tests/unit/models/test_assertions.py:78]:
```suggestion
    check.is_none(assertion.body)
```

File: tests/unit/models/test_assertions.py [hunk 5]
```diff
@@ -17,11 +17,92 @@
...
+                    ]
+                }
+            },
+        }
+    )
+
+    check.equal(assertion.body, None)
+    check.equal(assertion.body_length, None)
+    check.equal(assertion.sign_key_sha3_384, None)
+    check.equal(assertion.summary, None)
```
Comment (dariuszd21) [inline tests/unit/models/test_assertions.py:107]:
```suggestion
    check.is_none(assertion.body)
    check.is_none(assertion.body_length)
    check.is_none(assertion.sign_key_sha3_384)
    check.is_none(assertion.summary)
```

**Review by dariuszd21 — APPROVED** [review]
Looks good, thanks!

File: snapcraft/models/assertions.py (hunk 1 above)
Comment (lengau) [reply snapcraft/models/assertions.py:85]: Ooh good point/idea!

**Pull request merged**